util: Add a progress bar utility for gem5 simulation progress

erin-le opened the pull request:
This PR adds a utility that shows the progress of running gem5 simulations based on the current number of instructions executed compared to the total for a workload. It currently does not work with multisim and is not ready for review.

erin-le committed:
util: Add dashboard for viewing progress of gem5 simulations

This commit is an initial implementation of a dashboard util that shows
the user the progress of all gem5 processes on the current system.
It uses the tqdm library to add progress bars.

erin-le committed:
util: refactor gem5 dashboard progress bars

This commit creates a class to store a progress bar and data for
each progress bar. It refactors the code in gem5-dashboard.py to
use this class, allowing for a number of dictionaries to be
consolidated into one.

erin-le committed:
util: make modifications to avoid gem5 dashboard getting stuck

Under some circumstances, the gem5 dashboard won't correctly
exit or will get stuck when communicating with gem5 simulations
for stats. These changes attempt to prevent this.

erin-le committed:
util: minor changes to gem5 dashboard and orchestrator request

erin-le committed:
util: minor change to add error handling and fix issue

erin-le committed:
util: make the gem5 dashboard work with multisim

This change adds an additional check to prevent the dashboard
from sending SIGHUP (used to communicate with gem5 simulations)
to multisim processes that aren't gem5 simulations. Without this
change, the dashboard kills one of the multisim processes and
doesn't function correctly.

erin-le committed:
util: misc changes to gem5-dashboard.py

This commit modifies the gem5 dashboard so it doesn't print a
progress bar for simulations without total-instructions, instead
of exiting if any process is missing total-instructions. It also
increases the time between updates to the dashboard to 10 seconds,
from 0.1 seconds.

powerjg commented on src/python/gem5/simulate/simulator.py:
```diff
@@ -191,6 +191,18 @@ def __init__(
         self._last_exit_event = None
         self._exit_event_count = 0
 
+        # self._total_insts = int(self._board.get_workload()._estimated_instructions)
```
This doesn't need to be in the simulator. You can get the workload id in the simulator, then we can get information from there.

powerjg commented on src/sim/init_signals.cc:
```diff
@@ -451,6 +451,7 @@ initSignals()
     // Install a SIGIO handler to handle asynchronous file IO. See the
     // PollQueue class.
     installSignalHandler(SIGIO, ioHandler);
+    initSigRtmin();
```
Why this change?

erin-le committed:
util: Clean up comments, remove code in simulator.py

This commit removes some comments and also removes code in
simulator.py for returning the total_instructions from the
workload, as the total_instructions could be obtained from
the workload itself.

erin-le committed:
util: fix progress bar ordering when a process exits

This commit ensures that when a process exits, the progress bars
still stay in the same location and neatly display which processes
are still running and which have exited.

erin-le committed:
util: minor changes to gem5 dashboard and orchestrator request

erin-le committed:
sim: fix signal handler for hypercall external signal

This commit changes the signal handler for external gem5
orchestrator signals so that it matches the signal handlers
for other signals. It was previously set up incorrectly, which
occasionally caused seg faults when the hypercall_external_signal
utility was used to communicate with gem5.

erin-le committed:
sim, util: change SIGHUP to SIGCONT for shared mem transmitter

This commit changes the signal for the shared memory transmitter,
used by the gem5 orchestrator, from SIGHUP to SIGCONT. When a gem5
process is launched, there is a short period of time before signal
handlers are registered, during which the process will use the
signals' default behaviors. SIGHUP terminates processes by default,
which we want to avoid, so we instead use SIGCONT which is ignored
if a process isn't stopped.
This change fixes the issue of the hypercall_external_signal
utility killing gem5 simulations if it sends a signal too soon after
the simulations are launched.

erin-le committed:
util: change dashboard to display newly launched gem5 simulations

This commit changes the gem5 dashboard so that it shows
gem5 simulations that were launched after the dashboard was
launched.

erin-le committed:
python, sim, util: clean up comments and commented code

powerjg commented on src/sim/init_signals.cc:
```diff
@@ -453,9 +473,9 @@ initSignals()
 
-void initSigRtmin()
+void initSigCont()
 {
-    installSignalHandler(SIGHUP, externalProcessHandler);
+    installSignalHandler(SIGCONT, externalProcessHandler);
 }
```
Can you explain this change?

erin-le replied:
If the dashboard sends SIGHUP to a newly launched gem5 process that hasn't registered signal handlers yet, the process will use the default action for receiving SIGHUP and terminate. I changed the signal to SIGCONT, which is ignored by default.

This is mostly needed to add progress bars to the dashboard for simulations launched while the dashboard is running, though it could also prevent issues with launching the dashboard too soon after launching simulations.

powerjg commented on src/sim/simulate.cc:
```diff
@@ -327,6 +330,10 @@ doSimLoop(EventQueue *eventq)
                 async_exception = false;
                 return NULL;
             }
+            if (async_hypercall){
+                async_hypercall = false;
+                processExternalSignal();
+            }
```
Can you explain this change?

erin-le replied:
This is part of a change to the external signal handler that fixes segmentation faults caused by the dashboard. Previously, the external signal handler wasn't set up in the same way as the other signal handlers, so this changes the format to be the same. I'm not completely sure what was causing the seg faults, but it looks like fixing the format also fixed the seg faults.

powerjg commented on src/sim/simulate.cc:
```diff
@@ -191,13 +193,14 @@ simulate(Tick num_cycles)
     std::unique_ptr<GlobalSyncEvent, DescheduleDeleter> quantum_event;
 
-    inform("Entering event queue @ %d.  Starting simulation...\n", curTick());
+    DPRINTF(EnteringEventQueue, "Entering event queue @ %d. Starting "
+        "simulation...\n", curTick());
```
This is a good change, IMO, but it must be in another PR. This is a significant user-facing change that should have many eyes on it. Changes like this shouldn't be in the middle of a "util" PR.

erin-le replied:
Separate PR for this change opened here: https://github.com/gem5/gem5/pull/2215

powerjg commented on util/hypercall_external_signal/gem5-dashboard.py:
```diff
@@ -0,0 +1,131 @@
+import argparse
```
COPYRIGHT

powerjg commented on util/hypercall_external_signal/transmitter.py:
```diff
@@ -73,7 +73,7 @@
 from typing import Optional
 
-logging.basicConfig(level=logging.INFO)
+logging.basicConfig(level=logging.DEBUG)
```
You can change this on the command line instead of changing the hard-coded value.

powerjg commented on util/hypercall_external_signal/transmitter.py:
```diff
@@ -98,23 +101,32 @@ def send_signal(pid: int, id: int, payload: str) -> None:
-        shm.close()
-        shm.unlink()
-        sys.exit(1)
+        if "get_progress_bar_inst_count" in payload:
```
Rather than make this change, it would be better to factor out this function to send the exceptions to the caller. This script was not meant to be used by other scripts. It's not good practice to have sys.exit() in a function this way.

powerjg commented on util/hypercall_external_signal/orchestrator_request.py:
```diff
@@ -105,22 +108,63 @@ def find_gem5_pid() -> int:
+                                and "multiprocessing.resource_tracker"
+                                not in cmdline
+                            ):
+                                gem5_pids.append(int(pid))
```
This should be a function since it's the same for linux and mac

Also, you should avoid adding PIDs that are from other users since you can't send signals. Test each PID with a SIGCONT signal and see if you get a `PermissionError`

Finally, add some debug statements in here to facilitate debugging.

powerjg commented on src/python/gem5/simulate/exit_handler.py:
```diff
@@ -407,6 +407,18 @@ def _process(self, simulator: "Simulator") -> None:
                 elif function == "get_stats":
                     stats = simulator.get_stats()
```
I don't understand why this new code and new function is needed. Could you explain why you can't use `get_stats`? It has all of the information that it needed.

erin-le replied:
Originally, I wasn't sure if I should modify the existing functions, so I added `get_progress_bar_init_stats` to send the number of estimated total instructions + the other stats to initialize a progress bar. I was also planning to send signals to gem5 much more frequently, and thought that sending only the instruction count would improve performance, which isn't a significant concern anymore. I also ended up adding different timeout behavior for functions related to the dashboard. If it's better to use `get_stats`, I can try to refactor the code to use it and remove the new functions.

powerjg commented on src/python/gem5/resources/resource.py:
```diff
@@ -874,6 +874,15 @@ def __init__(
+        except:
+            self._estimated_instructions = None
+
+    def get_estimated_instructions(self) -> Optional[str]:
```
Shouldn't this return an int?

erin-le committed:
util: add ability to pass multiple pids to gem5 dashboard

erin-le committed:
sim: minor change to convert cerr to DPRINTF in init_signals.cc

erin-le committed:
sim, util: add macOS support to gem5 dashboard

erin-le committed:
util: partial refactor of gem5-dashboard.py (incomplete)

This contains a partial refactor of gem5-dashboard.py to improve
its structure based on changes suggested by Claude.

erin-le commented:
Closing in favor of https://github.com/gem5/gem5/pull/2250.
